src/libDirectoryService/PoWProcessing.cpp

```diff
@@ -556,25 +556,54 @@ void DirectoryService::RemoveReputationOfNodeFailToJoin(
 
 std::set<PubKey> DirectoryService::FindTopPriorityNodes(
     uint8_t& lowestPriority) {
+  LOG_MARKER();
+
   std::list<std::pair<PubKey, uint8_t>> listNodePriority;
   std::list<std::pair<PubKey, uint8_t>> listNewNodes;
+
+  auto size = MAX_SHARD_NODE_NUM < m_allPoWs.size() ? MAX_SHARD_NODE_NUM
+                                                    : m_allPoWs.size();
+
+  // maxShardGuards will be >= to trimmedGuardCount in SortPoWSoln
+  // But trimming in SortPoWSoln ensures there will be no more than
+  // SHARD_GUARD_TOL percent of shard guards
+  auto maxShardGuards = std::ceil(SHARD_GUARD_TOL * size);
+  uint32_t guardCounts = 0;
+
   // Iterate PoWs based on key ordering in the map
   for (const auto& kv : m_allPoWs) {
     const auto& pubKey = kv.first;
+    uint8_t priority = 0;
     auto reputation = m_mapNodeReputation.find(pubKey);
-    if (reputation != m_mapNodeReputation.end()) {
-      // listNodePriority is now ordered by key and contains only entries in
-      // m_allPoWs with reputation in m_mapNodeReputation
-      auto priority = CalculateNodePriority(reputation->second);
+
+    LOG_GENERAL(INFO, "Calculating priority for node: " << pubKey);
+
+    if (GUARD_MODE && Guard::GetInstance().IsNodeInShardGuardList(pubKey)) {
+      if (guardCounts >= maxShardGuards) {
+        LOG_GENERAL(INFO, "Enough shard guards, skipping " << pubKey);
+        continue;
+      }
+      priority = CalculateNodePriority(MAX_REPUTATION);
+      LOG_GENERAL(INFO, "Node " << pubKey
```

Comment:
For easier readability with the other logs below, suggest to change to `LOG_GENERAL(INFO, "Node=" << pubKey << " Reputation=(shard guard)");`